File: src/VL.DemoLib/04_DynamicEnum.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Collections.Specialized;
using System.Linq;
using System.Reactive.Linq;
using System.Text;
using System.Threading.Tasks;
using VL.Lib.Collections;
using VL.Core;

namespace DemoLib
{
    public class DynamicEnumDemo
    {
        public string Update(MyEnum enumInput)
        {
            if (enumInput.IsValid())
                return enumInput.Definition.Entries[enumInput.SelectedIndex()];
            else
                return "No valid entry selected";
        }

        public MyEnumDefinition GetDefinition()
        {
            return MyEnumDefinition.Instance;
        }
    }

    public class MyEnumDefinition : ManualDynamicEnumDefinitionBase<MyEnumDefinition>
    {
        //this is optional an can be used if any initialization before the call to GetEntries is needed
        protected override void Initialize()
        {
            //add two default entries on initialization
            AddEntry("abara", null);
            AddEntry("kadabara", null);
        }
    }

    [Serializable]
    public class MyEnum: DynamicEnumBase<MyEnum, MyEnumDefinition>
    {
        public MyEnum(string value) : base(value)
        {
        }

        //this method needs to be imported in VL to set the default
        public static MyEnum CreateDefault()
        {
            //use method of base class if nothing special required
            return CreateDefaultBase();
        }
    }
}

File: src/VL.FSharpDemoLib/04_DynamicEnum.fs
namespace FSharpDemoLib
open System
open System.Runtime.InteropServices //for the parameter attributes
open System.Collections.Generic
open System.Collections.Specialized
open System.Collections.ObjectModel
open VL.Lib.Collections
open System.Reactive.Linq

type MyEnumDefinition () = 
    inherit ManualDynamicEnumDefinitionBase<MyEnumDefinition> ()

    //add two default entries on initialization
    override this.Initialize() =
        this.AddEntry ("abara", null)
        this.AddEntry ("kadabara", null)
    
[<Serializable>]
type MyEnum (value:string) =
    inherit DynamicEnumBase<MyEnum, MyEnumDefinition> (value)

    //this method needs to be imported in VL to set the default
    static member CreateDefault() =
        //use method of base class if nothing special required
        DynamicEnumBase<MyEnum, MyEnumDefinition>.CreateDefaultBase()

type DynamicEnumDemo () =

    member this.Update (enumInput:MyEnum) = 
        if enumInput.IsValid()
        then enumInput.Definition.Entries.Item (enumInput.SelectedIndex())
        else "No valid entry selected"

    member this.AddEnumEntry entry = MyEnumDefinition.Instance.AddEntry (entry, null)
    member this.RemoveEnumEntry entry = MyEnumDefinition.Instance.RemoveEntry entry |> ignore

